Check if bank statement already imported

steveferry opened the pull request:
- Check every existing bank statement with the same profile as the one to import
- They are tested with an hash of the file, which is stored in the "Checksum" field if it doesn't exist

steveferry committed:
Check if bank statement already imported
Done with an hash of the file

yvaucher commented on l10n_ch_account_statement_base_import/statement.py:
```diff
@@ -71,8 +80,29 @@ def _statement_import(self, cr, uid, ids, prof, parser, file_stream,
+        """
+        Calculate hash of the file to determinate if it already exist
+        """
+        hash = hashlib.md5(file_stream).hexdigest()
```
`hash` is a buildin function please rename this variable

yvaucher commented on l10n_ch_account_statement_base_import/statement.py:
```diff
@@ -71,8 +80,29 @@ def _statement_import(self, cr, uid, ids, prof, parser, file_stream,
+        Calculate hash of the file to determinate if it already exist
+        """
+        hash = hashlib.md5(file_stream).hexdigest()
+        statement_obj = self.pool.get('account.bank.statement')
```
no need to use `get()` you can write simply `self.pool['account.bank.statement']`

yvaucher commented on l10n_ch_account_statement_base_import/statement.py:
```diff
@@ -71,8 +80,29 @@ def _statement_import(self, cr, uid, ids, prof, parser, file_stream,
+
+        for statement_id in statement_obj.search(
+                cr, uid, [('profile_id', '=', prof.id)], context=context):
+            statement = statement_obj.browse(cr, uid, statement_id,
```
It would be more efficient to browse all ids at once outside the for loop to reduce number of access to database.

yvaucher commented on l10n_ch_account_statement_base_import/statement.py:
```diff
@@ -26,6 +26,15 @@
+    _inherit = "account.bank.statement"
+
+    _columns = {
+        'checksum': fields.char(_('Checksum')),
```
No need to specify string as translatable. All field descriptions are already translatable

nbessi replied:
You should add index=True

yvaucher commented:
This is an interesting contrib thanks for it.

See my inline remarks plus check travis pylint remarks:

```
l10n_ch_account_statement_base_import/statement.py:85: [W0105(pointless-string-statement), AccountStatementProfil._statement_import] String statement has no effect
```

(you might want to add this in method docstring or change it as a comment with `#`

nbessi commented on l10n_ch_account_statement_base_import/statement.py:
```diff
@@ -71,8 +80,29 @@ def _statement_import(self, cr, uid, ids, prof, parser, file_stream,
+        statement_obj = self.pool.get('account.bank.statement')
+
+        for statement_id in statement_obj.search(
+                cr, uid, [('profile_id', '=', prof.id)], context=context):
```
Why don't directly do a search using hash I do not see the need of looping and browsing?

steveferry committed:
Search by checksum
As proposed by @nbessi, it's easier and a better way to search directly by checksum

ecino commented:
:+1:

yvaucher commented:
:+1:

nbessi commented:
:+1:

nbessi commented:
It would be nice to fix version 7.0 runbot before merging but it will not be blocking we have 3 thumbs

steveferry commented:
@nbessi  How can I fix it? I see nothing when i click on details

yvaucher commented:
Here is the error on runbot:

```
Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/cli/server.py", line 97, in preload_registry
    db, registry = openerp.pooler.get_db_and_pool(dbname,update_module=update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/pooler.py", line 33, in get_db_and_pool
    registry = RegistryManager.get(db_name, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/modules/registry.py", line 203, in get
    update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/modules/registry.py", line 233, in new
    openerp.modules.load_modules(registry.db, force_demo, status, update_module)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/modules/loading.py", line 329, in load_modules
    modobj.button_install(cr, SUPERUSER_ID, ids)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/addons/base/module/module.py", line 393, in button_install
    self.state_update(cr, uid, ids, 'to install', ['uninstalled'], context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/2948951-139-6720b1/openerp/addons/base/module/module.py", line 374, in state_update
    raise orm.except_orm(_('Error'), _("You try to install module '%s' that depends on module '%s'.\nBut the latter module is not available in your system.") % (module.name, dep.name,))
except_orm: (u'Error', u"You try to install module 'l10n_ch_lsv_dd' that depends on module 'account_direct_debit'.\nBut the latter module is not available in your system.")
```

We are trying to fix it in:

https://github.com/OCA/l10n-switzerland/pull/144

nbessi commented:
For me we can merge it is going to take some time to fix 7.0 runbot
